src/AggregateFunctions/AggregateFunctionTimeWeighted.h

```diff
@@ -0,0 +1,402 @@
+#pragma once
+
+#include <type_traits>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnsCommon.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <AggregateFunctions/AggregateFunctionSum.h>
+#include <Core/DecimalFunctions.h>
+#include <Common/logger_useful.h>
+#include <Poco/Logger.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+struct Settings;
+
+template <typename T> constexpr bool DecimalOrExtendedInt =
+    is_decimal<T>
+    || std::is_same_v<T, Int128>
+    || std::is_same_v<T, Int256>
+    || std::is_same_v<T, UInt128>
+    || std::is_same_v<T, UInt256>;
+
+/**
+ * Helper class to encapsulate values conversion for avg and avgWeighted.
+ */
+template <typename Numerator, typename Denominator>
+struct AvgTimeFraction
+{
+
+    struct Last
+    {
+        Numerator last_value;
+        Denominator last_time;
+    };
```

Comment:
```suggestion
    struct Last
    {
        Field last_value;
        Denominator last_time;
    };
```

Reply:
we just pass `value` to nested_func, so don't need care what is the type of value.

So only need a template parameter `TimeType` for the function